adam/flask_backend/app.py

```diff
@@ -135,7 +138,104 @@ def get_scene() -> Any:
     }
 
 
+@app.route("/api/get_image", methods=["GET"])
+def get_image() -> Any:
+    if not request.args:
+        abort(HTTPStatus.BAD_REQUEST)
+
+    image_path = request.args.get("image_path", default="")
+    with open(DATA_DIR / image_path, "rb") as fp:
+        image_data = b64encode(fp.read()).decode("utf-8")
+    return {"image_data": image_data}
+
+
+@app.route("/api/poll_changes", methods=["GET"])
+def poll_changes() -> Any:
+    if not request.args:
+        abort(HTTPStatus.BAD_REQUEST)
+    learner = request.args.get("learner", default="")
+    training_curriculum = request.args.get("training_curriculum", default="")
+    testing_curriculum = request.args.get("testing_curriculum", default="")
+    if not learner and not training_curriculum and not testing_curriculum:
+        abort(HTTPStatus.BAD_REQUEST)
+
+    curriculum_dir = (
+        LEARNERS_DIR
+        / learner
+        / EXPERIMENTS_DIR_NAME
+        / training_curriculum
+        / EXPERIMENTS_TESTING_DIR_NAME
+        / testing_curriculum
+    )
```

Comment:
I believe on most filesystems on unix-based devices, `/` is prohibited in file names, but there are no additional restrictions, so `.` can still appear. It's not advised and definitely bad practice, but I think we can be slightly looser on it. We can enforce that `/` does not appear in any of the input parameters, and then we can check that the resulting curriculum_dir is a relative file path to the LEARNERS_DIR using `curriculum_dir.is_relative_to(LEARNERS_DIR).

Reply:
But we can enforce no "." because they're not just filenames. They're filenames from a very limited set of possible names specifically defined by ADAM.

Reply:
Ah right, because the curriculum names are restriced based on what we manually allow ADAM to process. I thought for some reason the output directories could be separately configured in the parameter file.

Reply:
I see, that would make sense.

Reply:
In fact you're right, it is allowed to place the directories anywhere. I'm sorry. So I agree, forbid `/`s, permit `.`s as long as it resolves to the data directory.